project-koku/koku
# [COST-3617] - Fix GCP end of month crossover summary

## Jira Ticket

[COST-3617](https://issues.redhat.com/browse/COST-3617)

## Description

This change will drop invoice month from Trino SQL and run it per billing month based on start/end dates. This ensures we populate any crossover data for the previous month when running summary.

## Testing

The best way to test this is using the customer filtered data flow. If your not sure how to do this reach out to @lcouzens 

1. Checkout main
2. Restart Koku
3. Create a GCP storage only provider (using the Dev account + the lcouzens filtered data bucket)
4. Post the report already uploaded to that bucket `"custom-april-may-short.csv"`  
    1. This is done via the ingress reports api `/api/cost-management/v1/ingress/reports/`
    2. payload should look something like this: `{'source': '1', 'bill_year': '2024', 'bill_month': '05', 'reports_list': ['custom-april-may-short.csv']}`
5. Once ingested check trino tables for data compared to postgress
    1. TRINO: `select sum(cost), invoice_month from gcp_line_items_daily group by invoice_month order by invoice_month;`
    2. TRINO: `select sum(cost), invoice_month from gcp_line_items group by invoice_month order by invoice_month;`
    3. POSTGRES: `select sum(unblended_cost), invoice_month from reporting_gcpcostentrylineitem_daily_summary group by invoice_month order by invoice_month;`
    4. You should see that we only have data for a single invoice month
7. Repeat the above steps on this branch
8. Now you should see both months are correctly summarised in Trino/Postgres

We have since updated the local downloader too so its also possible to test in that flow similar to above:
1. Checkout main
2. Restart Koku
3. Generate Nise data for multiple months, combine these split month files into a single file (so we cross month bounds)
4. Put that combined file in your local dir and create the GCP-local provider that matches to ingest the data
5. Once ingested check trino tables for data compared to postgress
    1. TRINO: `select sum(cost), invoice_month from gcp_line_items_daily group by invoice_month order by invoice_month;`
    2. TRINO: `select sum(cost), invoice_month from gcp_line_items group by invoice_month order by invoice_month;`
    3. POSTGRES: `select sum(unblended_cost), invoice_month from reporting_gcpcostentrylineitem_daily_summary group by invoice_month order by invoice_month;`
    4. You should see that we only have data for a single invoice month
7. Repeat the above steps on this branch
8. Now you should see both months are correctly summarised in Trino/Postgres

## Release Notes

```markdown
* [COST-####](https://issues.redhat.com/browse/COST-####) Fix GCP end of month summary 
```
